# [AI 이노베이션 스퀘어] 11기 기본 저녁반
- github: https://github.com/hysKim1/AI_Fundamental
    - [파이썬 공식문서]https://docs.python.org/ko/3/
    - [Numpy 공식문서]https://numpy.org/devdocs/
    - [ML]https://developers.google.com/machine-learning/crash-course/ml-intro?hl=ko
    - [DL]https://www.tensorflow.org/
---

* [1. layers API ](#layers ) <br>
     *  [ weights](#weights)
     *  [variables](#variables)
     *  [ Layer Class 구성 ](#Layer-Class-구성)
         *  [\_\_init__](#\_\_init__)
         *  [\_\_call__](#\_\_call__)
         *  [build()](#build)
         *  [call()](#call)
             *  [built](#call)
             *  [build 따로 호츌한 경우](#build-따로-호츌한-경우)
             *  [build 없는 경우](#build-없는-경우)

             *  [응용 : Sequential model, Model 상속](#응용-:-Sequential-model)

* [2. 학습 방법 4가지 ](#학습-방법-4가지 ) <br>
     *  [2.1.train](#1.train)
     *  [2.2.fit](#2.fit)
         *  [batch](#batch)
         *  [epochs](#epochs)
         *  [verbose](#verbose)
         *  [callbacks](#callbacks)
         *  [validation_split](#validation_split)
         *  [class_weight](#class_weight)
         *  [initial_epoch](#initial_epoch)
         *  [steps_per_epoch](#steps_per_epoch)
         *  [workers](#workers)
         *  [trainable](#trainable)
     *  [2.3.train_on_batch](#3.train_on_batch)
          *  [train_step](#train_step)
     *  [2.4.GradientTape](#4.GradientTape)
---

## 모뎅 생성 방법 

1. tensor 이용 처음부터 끝까지 함 (low level)
    - 처음부터 끝까지 다 만듦
        - 지원하지 않는 기능을 만들어 사용할 때 씀
    - tensor 와 nn package
        - `tf.nn.relu`
2. estimator
    - 이미 만들어진 estimator 가져다가 씀

 keras 에서도 저수준으로 생성가능하여 권고함
 
3. Keras: `Sequential`
    - 단순,단방향
    - 편리
4. Keras:  `Model`
    1. functional paradigm
        - 이론,실제 간극을 줄임
    2. multi-input,multi-output 지원
        - 복잡한 모델을 만들 수 있음
    3. 미리 만들어진 기능 사용
    4. 유지 보수 쉬움

5. Keras: `Model` 상속(하위 클래스 sublass)
   -  만들어진 모델에 기능 추가
        - 상속해서 사용자 맞춤 레이어, 모델 사용 가능 
           - 예.`Callback` 상속
   - 동적 사용 가능(if, while 이용 조건에 따라서 바꿈)


- 혼재 사용되어 모든 것의 이해가 필요
- numpy array programmming , python 의 functional paradigm을 알아야 쉽게 이해 할 수 있음 
- 생성 모델마다 학습 방법이 조금씩 다름

In [1]:
import tensorflow as tf

#### Callback
- `Callbacks`:   모델 훈련 중 특정 지점에서 호출된 유틸리티
    - Callback을 상속받아 메소드를 overriding하면 customizing 가능

In [2]:
class MyCallback(tf.keras.callbacks.Callback):  #상속 예시 기억나시나요?
    pass 

## layers

- Keras layers API.


- `Model`, `Sequential`은 Input(데이터 크기), output layer(예측 갯수) 모양만 맞춰주면 알아서 자동으로 맞춰줌

-  `Model`, `Sequential` 사용하지 않기 때문에 `input_shape` 입력 없이도 에러 안나고 계산됨

In [3]:
# functional paradigm
layer= tf.keras.layers.Dense(100, input_shape=(3,)) #input layer
layer(tf.constant([[1,2,3]]))                       #

<tf.Tensor: shape=(1, 100), dtype=float32, numpy=
array([[-0.2830143 , -0.25614765,  0.8325263 ,  0.8277677 , -0.58371305,
         0.6336665 , -0.19517381,  0.6373619 , -0.00461085, -0.9215728 ,
        -0.5713495 , -0.61961097,  0.24371079,  0.84402585,  0.07088152,
         0.560688  ,  0.8192527 , -0.8428574 ,  0.00861928, -0.96730053,
         0.77552474,  0.030784  ,  0.08792338,  0.27490285,  0.19330627,
        -0.52598953,  0.21590787, -0.04829204,  0.04477295,  0.07227981,
         0.9533489 ,  0.15444304,  0.45035678,  1.0560806 , -0.4175315 ,
        -0.66157055,  0.36537766,  0.34576574, -0.8443349 , -0.18298252,
        -0.74625206,  0.47530472,  0.43119836, -0.04449588, -0.64954656,
        -0.71718025,  0.4823389 ,  0.83072436,  0.49941632, -0.31641358,
         0.26292595, -0.01962137, -0.05608597,  0.09870344, -0.4455146 ,
         0.79967487,  0.62288886, -0.07280165,  0.06239891,  0.27316946,
        -0.01538403, -0.97225606,  0.217953  ,  0.01639757, -0.5841864 ,
 

In [4]:
x=tf.constant([[1,2,3]])  #위와 동일
layer(x)                  #3개의 input 이 100개의 node에 각 곱해지게됨

<tf.Tensor: shape=(1, 100), dtype=float32, numpy=
array([[-0.2830143 , -0.25614765,  0.8325263 ,  0.8277677 , -0.58371305,
         0.6336665 , -0.19517381,  0.6373619 , -0.00461085, -0.9215728 ,
        -0.5713495 , -0.61961097,  0.24371079,  0.84402585,  0.07088152,
         0.560688  ,  0.8192527 , -0.8428574 ,  0.00861928, -0.96730053,
         0.77552474,  0.030784  ,  0.08792338,  0.27490285,  0.19330627,
        -0.52598953,  0.21590787, -0.04829204,  0.04477295,  0.07227981,
         0.9533489 ,  0.15444304,  0.45035678,  1.0560806 , -0.4175315 ,
        -0.66157055,  0.36537766,  0.34576574, -0.8443349 , -0.18298252,
        -0.74625206,  0.47530472,  0.43119836, -0.04449588, -0.64954656,
        -0.71718025,  0.4823389 ,  0.83072436,  0.49941632, -0.31641358,
         0.26292595, -0.01962137, -0.05608597,  0.09870344, -0.4455146 ,
         0.79967487,  0.62288886, -0.07280165,  0.06239891,  0.27316946,
        -0.01538403, -0.97225606,  0.217953  ,  0.01639757, -0.5841864 ,
 

In [5]:
layer 

In [6]:
dir(layer) #weigths, variables, bias, kernel...

['_TF_MODULE_IGNORED_PROPERTIES',
 '__call__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_activity_regularizer',
 '_add_trackable',
 '_add_variable_with_custom_getter',
 '_auto_track_sub_layers',
 '_autocast',
 '_autographed_call',
 '_batch_input_shape',
 '_build_input_shape',
 '_call_accepts_kwargs',
 '_call_arg_was_passed',
 '_call_fn_arg_defaults',
 '_call_fn_arg_positions',
 '_call_fn_args',
 '_call_full_argspec',
 '_callable_losses',
 '_cast_single_input',
 '_checkpoint_dependencies',
 '_clear_losses',
 '_compute_dtype',
 '_compute_dtype_object',
 '_dedup_weights',
 '_default_training_arg',
 '_deferred_dependencies',
 '_dty

### weights
layer 의 weigths 를 리스트 형태로 보여줌
keras의 `weight/variable` :bias, kernel


In [7]:
layer.weights

[<tf.Variable 'dense/kernel:0' shape=(3, 100) dtype=float32, numpy=
 array([[-0.09908906, -0.14917526,  0.09478635, -0.05901782, -0.15671882,
         -0.01969458,  0.14082211,  0.17910409, -0.10671273, -0.09651934,
         -0.18419048, -0.205289  , -0.08574831,  0.10794076, -0.23015988,
          0.12378696, -0.17909965,  0.09416717, -0.04186401, -0.05853239,
          0.18853888, -0.21989721,  0.23354033,  0.23351398,  0.07079718,
          0.06791946, -0.00263624,  0.04943761, -0.22357896, -0.06763031,
         -0.02471872, -0.07623239,  0.03207779,  0.02335292, -0.02387875,
          0.03121251, -0.121103  ,  0.14926967, -0.17954515, -0.0527703 ,
          0.09052438,  0.23333138, -0.06750469,  0.04842001, -0.02461134,
         -0.17067716, -0.12168367,  0.16591358,  0.15562353, -0.00961944,
          0.13735074,  0.14045537,  0.1464772 , -0.06257631, -0.17838255,
         -0.19194464,  0.02685538,  0.21107185,  0.18786481,  0.08462575,
         -0.01080593, -0.21156868,  0.182350

각 kernel, bias array를 unapcking 하여 나눠 저장

In [8]:
len(layer.weights) #2개 list

2

In [9]:
kernel,bias =layer.weights  #unpacking technique

In [10]:
kernel  #가중치  (3,100)

<tf.Variable 'dense/kernel:0' shape=(3, 100) dtype=float32, numpy=
array([[-0.09908906, -0.14917526,  0.09478635, -0.05901782, -0.15671882,
        -0.01969458,  0.14082211,  0.17910409, -0.10671273, -0.09651934,
        -0.18419048, -0.205289  , -0.08574831,  0.10794076, -0.23015988,
         0.12378696, -0.17909965,  0.09416717, -0.04186401, -0.05853239,
         0.18853888, -0.21989721,  0.23354033,  0.23351398,  0.07079718,
         0.06791946, -0.00263624,  0.04943761, -0.22357896, -0.06763031,
        -0.02471872, -0.07623239,  0.03207779,  0.02335292, -0.02387875,
         0.03121251, -0.121103  ,  0.14926967, -0.17954515, -0.0527703 ,
         0.09052438,  0.23333138, -0.06750469,  0.04842001, -0.02461134,
        -0.17067716, -0.12168367,  0.16591358,  0.15562353, -0.00961944,
         0.13735074,  0.14045537,  0.1464772 , -0.06257631, -0.17838255,
        -0.19194464,  0.02685538,  0.21107185,  0.18786481,  0.08462575,
        -0.01080593, -0.21156868,  0.18235081, -0.0931780

### variables
- `weights`가 이론에서 가리키는 가중치와 혼동되어 사용하는 별칭(alias)

In [11]:
layer.variables == layer.weights #둘이 동일 

True

In [12]:
layer.variables

[<tf.Variable 'dense/kernel:0' shape=(3, 100) dtype=float32, numpy=
 array([[-0.09908906, -0.14917526,  0.09478635, -0.05901782, -0.15671882,
         -0.01969458,  0.14082211,  0.17910409, -0.10671273, -0.09651934,
         -0.18419048, -0.205289  , -0.08574831,  0.10794076, -0.23015988,
          0.12378696, -0.17909965,  0.09416717, -0.04186401, -0.05853239,
          0.18853888, -0.21989721,  0.23354033,  0.23351398,  0.07079718,
          0.06791946, -0.00263624,  0.04943761, -0.22357896, -0.06763031,
         -0.02471872, -0.07623239,  0.03207779,  0.02335292, -0.02387875,
          0.03121251, -0.121103  ,  0.14926967, -0.17954515, -0.0527703 ,
          0.09052438,  0.23333138, -0.06750469,  0.04842001, -0.02461134,
         -0.17067716, -0.12168367,  0.16591358,  0.15562353, -0.00961944,
          0.13735074,  0.14045537,  0.1464772 , -0.06257631, -0.17838255,
         -0.19194464,  0.02685538,  0.21107185,  0.18786481,  0.08462575,
         -0.01080593, -0.21156868,  0.182350

### Layer Class 구성

> 실행순서
    0. init
        - layer에 필요한 매개변수를 입력
    1. __call__
        - build, call() 순서대로 호출
    2. build
        - 입력의 크기를 받고 나머지 초기화
        - 명시적으로 크기 지정할 필요없이 입력 바탕으로 자동으로 kernel 크기 맞춰 생성
        - __call__로 부터 한 번만 
    3. call()
        - forward compuation ( 정방향 연산) : 행렬 곱으로 feed forward함


### \_\_init__
- class 안에 정의하는 special/magic method `__  __`
- 인스턴스 생성하여 초기화하는 생성자

In [13]:
class X:
    def __init__(self):
        print('A')

In [14]:
X

__main__.X

In [15]:
X()

A


`super`:자식 클래스가 부모 클래스로부터 상속받은 멤버를 참조할 때 사용하는 참조 변수


    tf 공식문서 맞춤 레이어 참고: https://www.tensorflow.org/tutorials/customization/custom_layers?hl=ko
    
    
- tf.keras.Layer 클래스를 상속하고 하단처럼 구현

In [16]:
class MyDenseLayer(tf.keras.layers.Layer):
  def __init__(self, num_outputs):
    super().__init__()
    print('__init__')
    self.num_outputs = num_outputs

  def build(self, input_shape):
    print('build')
    print(input_shape)
    self.kernel = self.add_weight("kernel",
                                    shape=[int(input_shape[-1]),
                                           self.num_outputs])
    print(self.kernel.shape)

    
  def call(self, input):
    print('call')
    return tf.matmul(input, self.kernel)

4개의 output이며 아직 총 몇개의 데이터가 입력 받을지 모름
 ( None,4)

In [17]:
y=MyDenseLayer(4) #num_outputs =4

__init__


### \_\_call__
- 객체에 `()` 붙일 수 있음
- **클래스의 인스턴스를 callale하게 만들어 내부적으로 함수처럼 호출**
    - `build()` 존재하면 호출(optional)
        - self.kernel 초기화
        - 자동으로 input, output에 따라서  variables(kernel,bias) 크기 맞춰 줌
    - `call()` 호출

build를 정확히 언제 호출하는지 알아야함
 ### build
- optional
- 자동으로 input, output에 따라서  variables(kernel,bias) 크기 맞춰 줌
    - 행렬 곱  예: (l, m)@ (m, n) @ (n, y) @ (y a) = (l,a)
    
> 응용: Sequentials
   - 입력 레이어: `input_shape=(x,a)` 지정 해주면 알아서 맞춰 줌

> 응용:` Model ( inputs= ,ouputs= )`지정 해주면 알아서 맞춰 줌


In [18]:
y=MyDenseLayer(4) #num_outputs =4

__init__


1) build 호출 , input_shape=( 2,3)
    - self.kernel =(3,4) 자동으로 맞춰줌
2) call 호출
    - 행렬곱 : input @ self.kernel = (2,3)@(3,4)=(2,4) 

In [19]:
y(tf.zeros((2,3)))   #input_shape=(2,3)

build
(2, 3)
(3, 4)
call


<tf.Tensor: shape=(2, 4), dtype=float32, numpy=
array([[0., 0., 0., 0.],
       [0., 0., 0., 0.]], dtype=float32)>

In [20]:
y.kernel  #self.kernel (3,4) 생성됨

<tf.Variable 'my_dense_layer_1/kernel:0' shape=(3, 4) dtype=float32, numpy=
array([[-0.3817749 ,  0.10840201, -0.5969963 , -0.811849  ],
       [ 0.8055588 , -0.4150766 ,  0.22606778,  0.12713432],
       [-0.82967293,  0.38110697,  0.28664815,  0.44385839]],
      dtype=float32)>

### call
- `call()`을 직접적으로 부르면 `build()` 호출 안함

기존의 self.kernel = (3,4 )을 가져다 씀
-  input의 (행,열)에서 열을 3으로 맞추니 가능
- 행렬 곱 : (1,3)@(3,4)=(1,4)

In [21]:
y.call(tf.zeros((1,3)))  # 직접 call하면 build 안부르지만 kernel  크기에 맞춰 바꾸면 가능

call


<tf.Tensor: shape=(1, 4), dtype=float32, numpy=array([[0., 0., 0., 0.]], dtype=float32)>

InvalidArgumentError: Matrix size-incompatible: In[0]: [2,4], In[1]: [3,4] [Op:MatMul]
-  build를 부르지 않기 때문에 위에서 저장된 값 기반으로 실행
-  input=(2,4)와 kernel에서 나온 self.kernel (3,4) 행렬곱을 크기가 맞지 않아 할 수가 없음.

In [22]:
y.call(tf.zeros((2,4)))  # 직접 call하면 build 안 부른 경우 기존의 kenel 크기 안맞으면 에러 

call


InvalidArgumentError: Matrix size-incompatible: In[0]: [2,4], In[1]: [3,4] [Op:MatMul]

AttributeError: 'MyDenseLayer' object has no attribute 'kernel'
- build 부르지 않기 때문에 self.kernel 생성되지 않아 에러 남

In [23]:
y=MyDenseLayer(4)        #1) __init__  outputs=4
y.call(tf.zeros((2,4)))  #2) call()

__init__
call


AttributeError: 'MyDenseLayer' object has no attribute 'kernel'

In [24]:
y.kernel  #self.kernel 없기 때문에 없음

AttributeError: 'MyDenseLayer' object has no attribute 'kernel'

#### built
- `build()` 호축 되었는지 안 되었는지 여부를 boolean으로 반환값으로 확인
    - build 호출 됨:  x.built=True

In [25]:
x=MyDenseLayer(4)        #1) __init__ outputs=4
x.call(tf.zeros((2,4)))  #2) call()    input_shape ( 2,4) ->build 호출 안 함

__init__
call


AttributeError: 'MyDenseLayer' object has no attribute 'kernel'

In [26]:
x.built # build 호출 안됨 = False

False

In [27]:
x(tf.zeros((2,4)))   #build  호출  

build
(2, 4)
(4, 4)
call


<tf.Tensor: shape=(2, 4), dtype=float32, numpy=
array([[0., 0., 0., 0.],
       [0., 0., 0., 0.]], dtype=float32)>

`build` 는 인스턴스 한 다음 최초 한번만 호출됨
- 코드 출력값에 에 `build` 없음

In [28]:
x(tf.zeros((2,4)))   #build  호출 더이상 없음

call


<tf.Tensor: shape=(2, 4), dtype=float32, numpy=
array([[0., 0., 0., 0.],
       [0., 0., 0., 0.]], dtype=float32)>

In [29]:
x.built # build 호출 = True

True

In [30]:
x.built=False

In [31]:
x(tf.zeros((2,4)))   #build  다시 재호출  

build
(2, 4)
(4, 4)
call


<tf.Tensor: shape=(2, 4), dtype=float32, numpy=
array([[0., 0., 0., 0.],
       [0., 0., 0., 0.]], dtype=float32)>

#### build 따로 호츌한 경우
 - build에 입력된 값과 call 입력된값 일치시켜야 함

In [32]:
dd=MyDenseLayer(4)  #num_outputs=4

__init__


In [33]:
dd.build((3,2)) #build 부르면

build
(3, 2)
(2, 4)


In [34]:
dd.call(tf.ones((3,2))) #기존에 생성된 커널과 동일해야만 작동

call


<tf.Tensor: shape=(3, 4), dtype=float32, numpy=
array([[ 1.1630268 ,  0.4857781 ,  0.05285096, -0.6237049 ],
       [ 1.1630268 ,  0.4857781 ,  0.05285096, -0.6237049 ],
       [ 1.1630268 ,  0.4857781 ,  0.05285096, -0.6237049 ]],
      dtype=float32)>

In [35]:
dd.kernel

<tf.Variable 'kernel:0' shape=(2, 4) dtype=float32, numpy=
array([[ 0.30250072,  0.2723081 ,  0.4208634 ,  0.19015384],
       [ 0.8605261 ,  0.21346998, -0.36801243, -0.81385875]],
      dtype=float32)>

In [36]:
dd(tf.ones((1,9))) #build 재호출 되어 문제 없음

build
(1, 9)
(9, 4)
call


<tf.Tensor: shape=(1, 4), dtype=float32, numpy=
array([[-1.0523996 ,  0.6615981 , -0.5150342 ,  0.10197327]],
      dtype=float32)>

#### build 없는 경우
- 가능 하긴 하나 기본적으로 `Layer`, `Model` 클래스에서는 대부분 있으므로 상속해서 overriding 하려면 알아야함

In [37]:
class MyDenseLayer(tf.keras.layers.Layer):
  def __init__(self, num_outputs):
    super().__init__()
    print('__init__')
    self.num_outputs = num_outputs

  def call(self, input_):
    print('call')
    self.kernel = self.add_weight("kernel",
                                    shape=[int(input_.shape[-1]),
                                           self.num_outputs])
    print(self.kernel)
    return tf.matmul(input_, self.kernel)

In [38]:
b=MyDenseLayer(2)  

__init__


In [39]:
b.built                 #build 호출 없음

False

In [40]:
b.call(tf.ones((4,3))) 

call
<tf.Variable 'kernel:0' shape=(3, 2) dtype=float32, numpy=
array([[-0.0377878 , -0.88263446],
       [ 0.38724077, -1.0020505 ],
       [-0.3893581 ,  0.75553465]], dtype=float32)>


<tf.Tensor: shape=(4, 2), dtype=float32, numpy=
array([[-0.03990513, -1.1291504 ],
       [-0.03990513, -1.1291504 ],
       [-0.03990513, -1.1291504 ],
       [-0.03990513, -1.1291504 ]], dtype=float32)>

In [41]:
b.kernel #kernel 생성됨

<tf.Variable 'kernel:0' shape=(3, 2) dtype=float32, numpy=
array([[-0.0377878 , -0.88263446],
       [ 0.38724077, -1.0020505 ],
       [-0.3893581 ,  0.75553465]], dtype=float32)>

In [42]:
b(tf.ones((3,4))) # call : self.kernel에서 (6,3)생성 됨

call
<tf.Variable 'my_dense_layer_5/kernel:0' shape=(4, 2) dtype=float32, numpy=
array([[ 0.07729888, -0.04020834],
       [ 0.5306523 , -0.29695106],
       [ 0.6162038 ,  0.44001698],
       [ 0.2613976 , -0.20057702]], dtype=float32)>


<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
array([[ 1.4855525 , -0.09771943],
       [ 1.4855525 , -0.09771943],
       [ 1.4855525 , -0.09771943]], dtype=float32)>

In [43]:
class MyDenseLayer(tf.keras.layers.Layer):
  def __init__(self, num_outputs):
    super().__init__()
    print('__init__')
    self.num_outputs = num_outputs

  def call(self, input_):
    print('call')
    print(input_)
    return tf.matmul(input_, tf.ones((input_.shape[-1],self.num_outputs) ))
# 하단 부분 변경 

In [44]:
b=MyDenseLayer(2)    #build 없는 경우 

__init__


In [45]:
b.built

False

In [46]:
b(tf.ones((4,3))) #build 없어도 가능

call
tf.Tensor(
[[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]], shape=(4, 3), dtype=float32)


<tf.Tensor: shape=(4, 2), dtype=float32, numpy=
array([[3., 3.],
       [3., 3.],
       [3., 3.],
       [3., 3.]], dtype=float32)>

In [47]:
b.call(tf.ones((4,3))) #상동

call
tf.Tensor(
[[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]], shape=(4, 3), dtype=float32)


<tf.Tensor: shape=(4, 2), dtype=float32, numpy=
array([[3., 3.],
       [3., 3.],
       [3., 3.],
       [3., 3.]], dtype=float32)>

### 응용 : Sequential model
- functional paradigm 가능

In [48]:
#instance 방식
model= tf.keras.Sequential([ #input_shape없음 - summary  불가 
                           
        tf.keras.layers.Dense(2,activation='relu', name='layer1'),
        tf.keras.layers.Dense(3,activation='relu', name='layer2'),
        tf.keras.layers.Dense(4 ,name='layer3')
])

In [49]:
model.built #build 안 불림

False

`dir` build가  있는 것은 입력 크기 자동으로 맞춰주는 기능 갖춰주어 input,output 크기만 맞추면 됨

In [50]:
dir(model)      # __call__ ->callble->functional paradigm , build-> 자동 크기 맞춰줌

['_TF_MODULE_IGNORED_PROPERTIES',
 '__call__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_activity_regularizer',
 '_add_trackable',
 '_add_variable_with_custom_getter',
 '_assert_compile_was_called',
 '_assert_weights_created',
 '_auto_track_sub_layers',
 '_autocast',
 '_autographed_call',
 '_base_model_initialized',
 '_build_graph_network_for_inferred_shape',
 '_build_input_shape',
 '_call_accepts_kwargs',
 '_call_arg_was_passed',
 '_call_fn_arg_defaults',
 '_call_fn_arg_positions',
 '_call_fn_args',
 '_call_full_argspec',
 '_callable_losses',
 '_cast_single_input',
 '_check_call_args',
 '_checkpoint_dependencies',
 '_clear_los

In [51]:
x=tf.ones((3,5))  #input_shape= (3,5)
y=model(x)

In [52]:
model.built         #build 불림

True

코드를 기반으로 tensor 크기를 하단과 같이 알 수 있음
- input=(3,5)
- layer1 node 2 -> (5,2 )
- layer2 node 2 -> (2,2)
- layer3 node 4 -> output ( 2 , 4)

In [53]:
model(tf.ones((3,5))) #내부적으로 build를 부름

<tf.Tensor: shape=(3, 4), dtype=float32, numpy=
array([[ 0.29611617,  0.23772474,  0.2866837 , -0.24901707],
       [ 0.29611617,  0.23772474,  0.2866837 , -0.24901707],
       [ 0.29611617,  0.23772474,  0.2866837 , -0.24901707]],
      dtype=float32)>

In [54]:
model.summary() #위와 같은 방식으로 자동으로 맞춰주어 Output Shape 나옴

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
layer1 (Dense)               (3, 2)                    12        
_________________________________________________________________
layer2 (Dense)               (3, 3)                    9         
_________________________________________________________________
layer3 (Dense)               (3, 4)                    16        
Total params: 37
Trainable params: 37
Non-trainable params: 0
_________________________________________________________________


### 응용: 상속

#### 레이어 상속

- 단순하게 layer 묶은 것이기 때문애 모델안에 모델 들어갈 수 있음 
    - layer과 동일한 접근

In [55]:
#instance model
model= tf.keras.Sequential([ 
                           
        tf.keras.layers.Dense(2,activation='relu', name='layer1'),
        tf.keras.layers.Dense(2,activation='relu', name='layer2'),
        tf.keras.layers.Dense(4 ,name='layer3')
])

model1= tf.keras.Sequential([
            model,  # 위에거 가져옴
        tf.keras.layers.Dense(2,activation='relu', name='layer1')

])

In [56]:
model1.layers[0].layers # 단순히 묶음

#### Model 상속
- 상황에 따라 동적 프로그래밍 가능하여 복잡한 모델 생성 가능
- `Model` class 상속하는 고급 기법
    - \_\_init__  
        - flatten & dense 레이어 2개 
    - call
  
전문가를 위한 빠른 시작 공식 문서 참고:https://www.tensorflow.org/tutorials/quickstart/advanced?hl=ko

In [57]:
#https://www.tensorflow.org/tutorials/quickstart/advanced?hl=ko

class MyModel(tf.keras.models.Model):
  def __init__(self):
    super(MyModel, self).__init__()
    self.flatten = tf.keras.layers.Flatten()
    self.d1 = tf.keras.layers.Dense(128, activation='relu')
    self.d2 = tf.keras.layers.Dense(10, activation='softmax')

  def call(self, x):
    if x.shape[-1]>3 :                      #조건문에 따른 동적 프로그래밍 가능!
        x = self.flatten(x)
        x = self.d1(x)
        return self.d2(x)
    else:
        return self.d2(x)

In [58]:
model=MyModel() #안스턴스화

In [59]:
dir(model)     # __call__ ->callble->f

['_TF_MODULE_IGNORED_PROPERTIES',
 '__call__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_activity_regularizer',
 '_add_trackable',
 '_add_variable_with_custom_getter',
 '_assert_compile_was_called',
 '_assert_weights_created',
 '_auto_track_sub_layers',
 '_autocast',
 '_autographed_call',
 '_base_model_initialized',
 '_build_input_shape',
 '_call_accepts_kwargs',
 '_call_arg_was_passed',
 '_call_fn_arg_defaults',
 '_call_fn_arg_positions',
 '_call_fn_args',
 '_call_full_argspec',
 '_callable_losses',
 '_cast_single_input',
 '_check_call_args',
 '_checkpoint_dependencies',
 '_clear_losses',
 '_compile_was_called',
 '_compiled_tr

In [60]:
model(tf.ones(10,5))  # _call_ -> build(),call()

<tf.Tensor: shape=(10, 10), dtype=float32, numpy=
array([[0.08784313, 0.10958985, 0.09111308, 0.12228402, 0.11307134,
        0.09901255, 0.10082204, 0.09111098, 0.08669054, 0.09846248],
       [0.08784313, 0.10958985, 0.09111308, 0.12228402, 0.11307134,
        0.09901255, 0.10082204, 0.09111098, 0.08669054, 0.09846248],
       [0.08784313, 0.10958985, 0.09111308, 0.12228402, 0.11307134,
        0.09901255, 0.10082204, 0.09111098, 0.08669054, 0.09846248],
       [0.08784313, 0.10958985, 0.09111308, 0.12228402, 0.11307134,
        0.09901255, 0.10082204, 0.09111098, 0.08669054, 0.09846248],
       [0.08784313, 0.10958985, 0.09111308, 0.12228402, 0.11307134,
        0.09901255, 0.10082204, 0.09111098, 0.08669054, 0.09846248],
       [0.08784313, 0.10958985, 0.09111308, 0.12228402, 0.11307134,
        0.09901255, 0.10082204, 0.09111098, 0.08669054, 0.09846248],
       [0.08784313, 0.10958985, 0.09111308, 0.12228402, 0.11307134,
        0.09901255, 0.10082204, 0.09111098, 0.08669054, 0.09

 ---
 초보자를 위한 시작 https://www.tensorflow.org/tutorials/quickstart/beginner?hl=ko

In [113]:
#초보자를 위한 시작 https://www.tensorflow.org/tutorials/quickstart/beginner?hl=ko

mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

In [62]:
#Sequential
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10, activation='softmax')
])

In [63]:
# Model - 상동
inputs=tf.keras.Input(shape=(28,28))
flatten=tf.keras.layers.Flatten()(inputs)

layer1=tf.keras.layers.Dense(128,activation='relu')(flatten)
dropout=  tf.keras.layers.Dropout(0.2)(layer1)
output= tf.keras.layers.Dense(10, activation='softmax')(dropout)

model1=tf.keras.models.Model(inputs=inputs,outputs=output)

In [64]:
model.built

True

In [65]:
model.summary()  #모델 생성 완료

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_1 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_3 (Dense)              (None, 128)               100480    
_________________________________________________________________
dropout (Dropout)            (None, 128)               0         
_________________________________________________________________
dense_4 (Dense)              (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________


#### add_ method

#### add_loss
- compile( loss=) 에서 제공하는 것 외에 loss function 을 사용하고 싶을때 사용
- subclassed layer or model의 상속을 받는 하위 클래스 내부에 'call' 함수에서 사용 가능

stackoverflow 참고 :https://stackoverflow.com/questions/50063613/what-is-the-purpose-of-the-add-loss-function-in-keras

- `SparseCategoricalCrossentropy`: label encoding 다중 분류문제에 사용

In [66]:
model.add_loss(tf.keras.losses.SparseCategoricalCrossentropy())

add_loss loss가 2개 돌아가기 때문에 한 번 정하면 compile 해도 변경이 안됨
- 사용자 정의 학습 할때 사용
- compile -> fit

In [67]:
model.compile(loss=tf.keras.losses.SparseCategoricalCrossentropy())

In [68]:
model.fit(x_train,y_train,epochs=5)

Epoch 1/5


TypeError: in user code:

    /Users/MG/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py:806 train_function  *
        return step_function(self, iterator)
    /Users/MG/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py:796 step_function  **
        outputs = model.distribute_strategy.run(run_step, args=(data,))
    /Users/MG/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/distribute/distribute_lib.py:1211 run
        return self._extended.call_for_each_replica(fn, args=args, kwargs=kwargs)
    /Users/MG/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/distribute/distribute_lib.py:2585 call_for_each_replica
        return self._call_for_each_replica(fn, args, kwargs)
    /Users/MG/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/distribute/distribute_lib.py:2945 _call_for_each_replica
        return fn(*args, **kwargs)
    /Users/MG/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py:789 run_step  **
        outputs = model.train_step(data)
    /Users/MG/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py:749 train_step
        y, y_pred, sample_weight, regularization_losses=self.losses)
    /Users/MG/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/keras/engine/base_layer.py:1433 losses
        loss_tensor = regularizer()
    /Users/MG/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/keras/engine/base_layer.py:1509 _tag_callable
        loss = loss()

    TypeError: __call__() missing 2 required positional arguments: 'y_true' and 'y_pred'


In [69]:
#한 가지 예시로 
inputs = tf.keras.Input(shape=(28,28))
x = tf.keras.layers.Dense(10)(inputs)
outputs = tf.keras.layers.Dense(1)(x)
modell = tf.keras.Model(inputs, outputs)

# Activity regularization- 종속성이 다른 경우

modell.add_loss(tf.abs(tf.reduce_mean(x)))

In [70]:
modell.compile(loss='sparse_categorical_crossentropy', 
              optimizer='adam')#- fit 사용하기 위해서 compile 해야함`

In [71]:
modell.fit(x_train,y_train,epochs=5)  #activity regularization 으로 가능

Epoch 1/5
1875/1875 [==============================] - 3s 2ms/step - loss: 3.5449
Epoch 2/5
1875/1875 [==============================] - 2s 1ms/step - loss: 3.2854
Epoch 3/5
1875/1875 [==============================] - 3s 2ms/step - loss: 3.2680
Epoch 4/5
1875/1875 [==============================] - 2s 1ms/step - loss: 3.2662
Epoch 5/5
1875/1875 [==============================] - 2s 1ms/step - loss: 3.2670


- 보통 `compile (loss, metrics)`에서 제공되지 않는 경우
- 활성화만 요구할 수 있으며(이 경우에는 대상이 없음), 이 활성화는 모델 출력이 아닐 수 있다.

 - ActivityRegularizationLayer 경우에는 activation of a layer만 필요함(target 불필요) 
 - 커스텀한 layer의 call method 로 self.add_loss(loss_value)를 호출 가능.
- 이러한 방식으로 추가된 손실을 학습 동안에 `compile`로 보내어지는 loss에 추가된다. 

TF 공식문서 학습과 평가 참고 : https://www.tensorflow.org/guide/keras/train_and_evaluate

## 학습 방법 4가지
상황에따라서 학습 방법이 다름
- 모델 만드는법 5가지 : tensor , estimator, keras(Sequentials, Model 함수형 패러다임, Model subclass)
- 학습 방법 4가지: train,fit, train_on_batch, GradientTape

혼용 주의

## 1.train
- estimator는 scikit-learn이랑 유사.
    - compile 없이 바로 학습 
    - 둘 다 한번에 layer,node 갯수 정함
    - 상세한 변경은 불가하여 맞춤 학습이 어려움

Init signature:
MLPClassifier(
    hidden_layer_sizes=(100,),
    activation='relu',
    solver='adam',
    alpha=0.0001,
    batch_size='auto',
    learning_rate='constant',
    learning_rate_init=0.001,
    power_t=0.5,
    max_iter=200,
    shuffle=True,
    random_state=None,
    tol=0.0001,
    verbose=False,
    warm_start=False,
    momentum=0.9,
    nesterovs_momentum=True,
    early_stopping=False,
    validation_fraction=0.1,
    beta_1=0.9,
    beta_2=0.999,
    epsilon=1e-08,
    n_iter_no_change=10,
    max_fun=15000,
)

## 2.fit
 - 상황에 맞게 조절하여 사용 
 
### batch
- 학습 시 몇 개마다 업데이트 할 것인지
    - 크기가 클수록 빠르나 정확도가 떨어짐 
    - 최소 1 ~ 최대 전체 샘플 갯수
    - 비유:수학 문제를 다 풀고 답을 풀것인지 나눠서 확인하고 다 풀것인지..

### epochs 
- 데이터 전체를 몇번 학습 시킬지

In [73]:
dir(model1) #fit, train_ ...

['_TF_MODULE_IGNORED_PROPERTIES',
 '__call__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_activity_regularizer',
 '_add_trackable',
 '_add_variable_with_custom_getter',
 '_assert_compile_was_called',
 '_assert_weights_created',
 '_auto_track_sub_layers',
 '_autocast',
 '_autographed_call',
 '_base_model_initialized',
 '_build_input_shape',
 '_call_accepts_kwargs',
 '_call_arg_was_passed',
 '_call_fn_arg_defaults',
 '_call_fn_arg_positions',
 '_call_fn_args',
 '_call_full_argspec',
 '_callable_losses',
 '_cast_single_input',
 '_check_call_args',
 '_checkpoint_dependencies',
 '_clear_losses',
 '_compile_was_called',
 '_compiled_tr

### verbose 
- 설명 양 
- verbose: Integer. 0, 1, or 2(- 숫자카 커질수록 설명 양이 적어짐)
> Verbosity mode. 
   - 0 = silent, 
   - 1 = progress bar, 
   - 2 = one line per epoch
   - 3 >= epoch

In [75]:
model1.compile(loss='sparse_categorical_crossentropy', 
              optimizer='adam') 
model1.fit(x_train,y_train,epochs=3, verbose=0)

In [77]:
model1.fit(x_train,y_train,epochs=3, verbose=1)

Epoch 1/3
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0448
Epoch 2/3
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0414
Epoch 3/3
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0377


In [78]:
model1.fit(x_train,y_train,epochs=3, verbose=2)

Epoch 1/3
1875/1875 - 7s - loss: 0.0350
Epoch 2/3
1875/1875 - 6s - loss: 0.0323
Epoch 3/3
1875/1875 - 6s - loss: 0.0323


In [79]:
model1.fit(x_train,y_train,epochs=3, verbose=4)

Epoch 1/3
Epoch 2/3
Epoch 3/3


### callbacks
- 한 번 실행하고 나서 epoch 실행되게 함
- 학습과 검증 과정에서 적용할 keras.callbacks.Callback 인스턴스의 리스트.
    -  학습 과정의 특정 단계에서 적용할 함수의 세트
        - keras.callbacks.Callback :확작해서 커스텀 콜백 생성 가능
        - keras.callbacks.BaseLogger
        - keras.callbacks.TerminateOnNaN
        - keras.callbacks.ProgbarLogger
        - keras.callbacks.History
        - keras.callbacks.ModelCheckpoint
        - keras.callbacks.RemoteMonitor
        - LearningRateScheduler
        - keras.callbacks.TensorBoard
        - keras.callbacks.ReduceLROnPlateau
        - keras.callbacks.CSVLogger

keras 공식 문서 참고 : https://keras.io/ko/callbacks/#_1


In [ ]:
# 각 batch에 대한 loss  내역 기록
class LossHistory(keras.callbacks.Callback): #상속 -> self.model  접근 가능
    def on_train_begin(self, logs={}):
        self.losses = []

    def on_batch_end(self, batch, logs={}):
        self.losses.append(logs.get('loss'))

        
model = Sequential()
model.add(Dense(10, input_dim=784, kernel_initializer='uniform'))
model.add(Activation('softmax'))
model.compile(loss='categorical_crossentropy', optimizer='rmsprop')

history = LossHistory()  #인스턴스화 한 후 fit에 사용
model.fit(x_train, y_train, batch_size=128, epochs=20, verbose=0, callbacks=[history])

print(history.losses)

In [ ]:
# epoch마다 모델 저장
from keras.callbacks import ModelCheckpoint

model = Sequential()
model.add(Dense(10, input_dim=784, kernel_initializer='uniform'))
model.add(Activation('softmax'))
model.compile(loss='categorical_crossentropy', optimizer='rmsprop')

'''
만약 검증에 대하 loss가 감소하다면 각 epoch  후의 모델의 가중치를 저장함
'''
checkpointer = ModelCheckpoint(filepath='/tmp/weights.hdf5', verbose=1, save_best_only=True)
model.fit(x_train, y_train, batch_size=128, epochs=20, verbose=0, validation_data=(X_test, Y_test), callbacks=[checkpointer])

### validation_split
- 내부적으로 hold out으로 기본 0.2는 전체의 80%는 학습 나머지 20%은 성능검정에 사용해서 매 epoch 마다 val_loss 출력

In [251]:
model1.fit(x_train,y_train,epochs=3, validation_split=0.2)  #val_loss 생김

Epoch 1/3
1500/1500 [==============================] - 9s 6ms/step - loss: 0.0840 - val_loss: 0.0638
Epoch 2/3
1500/1500 [==============================] - 8s 6ms/step - loss: 0.0797 - val_loss: 0.0690
Epoch 3/3
1500/1500 [==============================] - 8s 5ms/step - loss: 0.0766 - val_loss: 0.0681


predict 단순히 input 을 넣어서 예측값 넣음
차이를 줄이는것이 loss 함수이며 줄여가는 과정을 학습
- back propagation 
- 뒤로가면서 loss가 가장 작은 weight값 찾음

### class_weight
- imbalanced dataset : 데이터가 불균형한 경우 특정 데이터들은 잘 학습이 안되기 때문에 데이터들에 가중치를 더 주어 집중 학습
    - [1,1,1,1,1,1,1,1,0,0] 데이터에서 0은 수가 절대적으로 적어서 학습이 잘 안됨

### sample_weight
- 전문가,초보자가 수집한 데이터중 신뢰성이 다르기 때문에 데이터 별로 가중치를 신뢰도에 따라 다르게 줌

### initial_epoch
- 학습 시작시 몇번째 epcoch부터 시작할지 정하여 건너 뛰고 학습
- 학습시킬때마다 누적되기 때문에 학습을 시작하는 epcoh (이전의 학습 과정을 재개하는데 유용)

In [253]:
model1.fit(x_train,y_train,epochs=3, initial_epoch=1)

Epoch 2/3
1875/1875 [==============================] - 7s 4ms/step - loss: 0.0816
Epoch 3/3
1875/1875 [==============================] - 7s 4ms/step - loss: 0.0757


In [82]:
model1.fit(x_train,y_train,epochs=3, initial_epoch=2)

Epoch 3/3
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0239


### steps_per_epoch
- step: Weight와 Bias를 1회 업데이트
 $steps_per_epoch=\frac {no samples X epochs}{batch_size}$
- 전체 데이터 하나(1 epoch)에 대해서 weight,Bias를 업데이트 하는 횟수(총 step횟수=batches of samples)

In [104]:
model1.fit(x_train,y_train,epochs=3, steps_per_epoch=3)

Epoch 1/3
3/3 [==============================] - 0s 140ms/step - loss: 0.0344
Epoch 2/3
3/3 [==============================] - 0s 111ms/step - loss: 0.0321
Epoch 3/3
3/3 [==============================] - 0s 99ms/step - loss: 0.0346


In [100]:
model1.fit(x_train,y_train,epochs=3,batch_size=60000) #상동

Epoch 1/3
1/1 [==============================] - 0s 1ms/step - loss: 0.0357
Epoch 2/3
1/1 [==============================] - 0s 2ms/step - loss: 0.0351
Epoch 3/3
1/1 [==============================] - 0s 2ms/step - loss: 0.0345


계산이 안 맞는경우 다음과같이 됨.
- `steps_per_epoch * epochs` batches 갯수까지 가능한데 초과 하므로 학습 경고메시지 생김
    - steps_per_epoch 먼저 맞춤 
- 4 *3 =12 개의 batches 가능하나 최대 batch_size= 60,000/12=5,000개


In [102]:
model1.fit(x_train,y_train,epochs=3,batch_size=60000,steps_per_epoch=4) #경고

Epoch 1/3
3/4 [=====================>........] - 1s 291ms/step - loss: 0.0339


In [103]:
model1.fit(x_train,y_train,epochs=3,batch_size=5000,steps_per_epoch=4) #잘 됨

Epoch 1/3
4/4 [==============================] - 0s 45ms/step - loss: 0.0352
Epoch 2/3
4/4 [==============================] - 0s 48ms/step - loss: 0.0338
Epoch 3/3
4/4 [==============================] - 0s 48ms/step - loss: 0.0316


### workers
- scikit-learn의 `n_jobs`: 동시에 core 몇개 사용할지
- 가속을 위한 프로세스의 최대 갯수를 정수형으로 입력 

In [107]:
model1.fit(x_train,y_train,epochs=3, workers=5)

Epoch 1/3
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0204
Epoch 2/3
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0187
Epoch 3/3
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0213


In [ ]:
dir(model1) #train_step

### trainable
- 독립적으로 학습 시킬지 아닐지 여부를 정함
- 'freezing' 이라 불리우며 `fit`학습 시킬때 weights 업데이트 시키지 않고 그대로 가져오기 때문

참고: https://www.tensorflow.org/guide/keras/transfer_learning

In [40]:
inputs=tf.keras.Input(shape=(28,28))
flatten=tf.keras.layers.Flatten()(inputs)
layer1=tf.keras.layers.Dense(128,activation='relu')(flatten)

layer1.trainable=False                            # 추가 frozen layer1

dropout=  tf.keras.layers.Dropout(0.2)(layer1)
output= tf.keras.layers.Dense(10, activation='softmax')(dropout)
model1=tf.keras.models.Model(inputs=inputs,outputs=output)


model1.compile(loss='sparse_categorical_crossentropy')

- 육안으로 변화 없음 trainable에대한 summary 버그라는 글 찾음: https://github.com/keras-team/keras/issues/4674


- 학습시킬수 있는 파라미터수 : Trainable params: 101,770


In [39]:
model1.summary() 

Model: "functional_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_1 (InputLayer)         [(None, 28, 28)]          0         
_________________________________________________________________
flatten (Flatten)            (None, 784)               0         
_________________________________________________________________
dense_6 (Dense)              (None, 128)               100480    
_________________________________________________________________
dropout (Dropout)            (None, 128)               0         
_________________________________________________________________
dense_7 (Dense)              (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________


In [35]:
model1.layers  # 모든 레이어 속함

In [36]:
model1.layers[-1].weights  #ㅐoutput layer 

[<tf.Variable 'dense_12/kernel:0' shape=(128, 10) dtype=float32, numpy=
 array([[-0.07351604, -0.18875545,  0.17297213, ...,  0.14054741,
         -0.0016579 , -0.16374114],
        [-0.03305444,  0.03852634, -0.08335005, ...,  0.04514392,
         -0.20475346,  0.00587885],
        [-0.19760737, -0.20600028,  0.09294061, ...,  0.17326348,
          0.05185036, -0.04805392],
        ...,
        [ 0.01094228,  0.07453905,  0.06103332, ...,  0.11846419,
          0.11747263, -0.14852801],
        [-0.14051121, -0.01900749,  0.17888589, ...,  0.11291544,
          0.09285702, -0.03292727],
        [ 0.05307014,  0.09761281, -0.10868098, ...,  0.19911365,
          0.07572933, -0.1559089 ]], dtype=float32)>,
 <tf.Variable 'dense_12/bias:0' shape=(10,) dtype=float32, numpy=array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.], dtype=float32)>]

 Non-trainable params: 0 에서 1,290 이 됨.

In [44]:
model1.layers[-1].trainable=False  #freezing output layer 
model1.summary()

Model: "functional_9"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_5 (InputLayer)         [(None, 28, 28)]          0         
_________________________________________________________________
flatten_4 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_11 (Dense)             (None, 128)               100480    
_________________________________________________________________
dropout_4 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_12 (Dense)             (None, 10)                1290      
Total params: 101,770
Trainable params: 100,480
Non-trainable params: 1,290
_________________________________________________________________


In [45]:
model1.layers[-1].weights  #안변함

[<tf.Variable 'dense_12/kernel:0' shape=(128, 10) dtype=float32, numpy=
 array([[-0.07351604, -0.18875545,  0.17297213, ...,  0.14054741,
         -0.0016579 , -0.16374114],
        [-0.03305444,  0.03852634, -0.08335005, ...,  0.04514392,
         -0.20475346,  0.00587885],
        [-0.19760737, -0.20600028,  0.09294061, ...,  0.17326348,
          0.05185036, -0.04805392],
        ...,
        [ 0.01094228,  0.07453905,  0.06103332, ...,  0.11846419,
          0.11747263, -0.14852801],
        [-0.14051121, -0.01900749,  0.17888589, ...,  0.11291544,
          0.09285702, -0.03292727],
        [ 0.05307014,  0.09761281, -0.10868098, ...,  0.19911365,
          0.07572933, -0.1559089 ]], dtype=float32)>,
 <tf.Variable 'dense_12/bias:0' shape=(10,) dtype=float32, numpy=array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.], dtype=float32)>]

모델에 적용하면 해당 모델에 속한 레이어 전체에 적용됨
- Non-trainable params: 101,770 전체

In [48]:
model1.trainable=False  #모든 레이어 
model1.summary()  

Model: "functional_9"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_5 (InputLayer)         [(None, 28, 28)]          0         
_________________________________________________________________
flatten_4 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_11 (Dense)             (None, 128)               100480    
_________________________________________________________________
dropout_4 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_12 (Dense)             (None, 10)                1290      
Total params: 101,770
Trainable params: 0
Non-trainable params: 101,770
_________________________________________________________________


Runs a single gradient update on a single batch of data.
1의 batch 데이터에서 업데이트

In [49]:
model1.fit(x_train,y_train,epochs=3)

Epoch 1/3
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0297A: 0s -
Epoch 2/3
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0274
Epoch 3/3
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0254


In [50]:
model.weights #초기 가중치 가져옴

[<tf.Variable 'dense_2/kernel:0' shape=(784, 128) dtype=float32, numpy=
 array([[ 0.04118228,  0.04027221, -0.05904735, ..., -0.07277236,
         -0.0618301 ,  0.06054585],
        [-0.07615273, -0.04857934, -0.0197164 , ..., -0.06975673,
          0.07392184, -0.02188018],
        [ 0.04434947,  0.0651508 ,  0.01847351, ...,  0.03339616,
         -0.04772325,  0.07733341],
        ...,
        [ 0.00349327, -0.06029075, -0.04563036, ...,  0.04359804,
         -0.073309  , -0.02422466],
        [-0.04939568, -0.04466948,  0.05060906, ...,  0.02847371,
         -0.0441247 , -0.00334771],
        [-0.07694268,  0.00519963,  0.01613969, ...,  0.01255116,
         -0.01917897,  0.01981416]], dtype=float32)>,
 <tf.Variable 'dense_2/bias:0' shape=(128,) dtype=float32, numpy=
 array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,

## 3.train_on_batch
- 1 batch 당 1번의 gradient업데이트로 얻은 training loss 
- fit의 저수준 버전

In [273]:
model1.train_on_batch(x_train,y_train)  #fit과 비슷

2.540980100631714

- fit은 하단을 train_on_batch for 이용하여 array연산  
    - 4 levels : train > fit > **train_on_batch** > GradientTape

In [275]:
for i in range(3):
    print(model1.train_on_batch(x_train,y_train))

0.41024714708328247
0.405737042427063
0.3974108397960663


#### train_step 
  - `fit()`에서 반복적 호출( tf.version 2.2 이상 부터 )하여 사용자가 정의한 학습 loop을 조정 가능함
  https://www.tensorflow.org/api_docs/python/tf/keras/Model

## 자동 미분
- 학습: gradient descent 통해  loss function 최소화하는 파라미터 찾음
    - loss function : 입력 & 실체값 차이를 함수로 표현
    - gradient descent : 현재 위치와 learning rate곱하여 이동하면서 loss/cost function 최소값(미분값=0)을 주는 파라미터를 찾음
        - gradient : 특정점의 기울기
        - learning rate $\alpha$: 현재 그래디언트에서 다음 위치까지 얼마만큼 가야하는지
        - 미분 : 하나의 변수      
        - 편미분: 변수가 여러개 인 경우 각각에 대해 미분)
            - 자동 미분 :auto differentiation
                - tensorflow  지원 4가지: `train`,`fit`, `train_on_batch`, `GradientTape`

## 4.GradientTape
- 1970년대에 인기 저장 매체로 tape 사용하여 영상물 기록
- 주어진 입력에 대한 연산의 gradient에서 최소점을 찾기위해 학습한 것이 어떻겨 변화하는지 편미분값 저장
- 가장 저주순
- 미분하기위해서 미분하는 함수를 GradientTape에 넣으면 자동미분 할 수 있도록 구조체로 변경
    - 변수가 여러개 이므로 편미분을 통해서함 

tf  공식문서 자동미분과 그래디어트 테이프 참고 : https://www.tensorflow.org/tutorials/customization/autodiff?hl=ko
- 자세한것은 다음 시간에 이어서